### PR TITLE
Feature: add more git buildstep (git add and git commit)

### PR DESCRIPTION
**Pull request by chapipo**
This PR add two buildsteps:
- GitAdd to add some files in the staging area
- GitCommit to commit the files in the staging area

I have removed old styles docs from the source (like describe in the docs) and add some trivial test cases in other git related steps.

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by p12tic** [comment]
Thanks for the PR! 

What do you think if there was a single step to add files and commit, i.e. `GitCommit` would accept a list of files and would do both `git add` and `git commit`? Did you have a use case where `GitAdd` does not come with a paired `GitCommit`?

**Comment by chapipo** [comment]
When I start the work on this, I ask myself the same question !

I don't have use case where `GitAdd` step do not come with `GitCommit` step. It's more to keep it close to git command line and have one git command by step.

But I can rework the PR to do only one step if you tell me so ;)

**Comment by p12tic** [comment]
Buildbot does not necessarily need to map command line 1:1. It would be easier for the end users to configure only single step, not two and also reduce the chance of accidental errors. In addition to that, the staging area in git is pretty unique capability in version control systems so not exposing it would make the interface of the step more similar to e.g. HgCommit or SvnCommit once we have them.

I'll review the current PR to look for any problems that are likely to persist after the refactor, so don't hurry in rewriting it :-) Thanks!

**Comment by tardyp** [comment]
I agree with @p12tic and I would even go further.

We should have have a GitPush command, which can do the git branch, git add, git commit, git tag, and git push in one step.
With a nice API so that it is easy to do all those things.

**Comment by p12tic** [comment]
> We should have have a GitPush command, which can do the git branch, git add, git commit, git tag, and git push in one step.
I think that users may want to do multiple commits before pushing, that would be hard in such an API. Push is an atomic event, so multiple `GitPush` may not work in some cases.

**Comment by tardyp** [comment]
> I think that users may want to do multiple commits before pushing,

okay

**Comment by chapipo** [comment]
I have updated the code and the tests.

@p12tic it will be easier to review. A lot of code as duplicated with 2 steps instead of 1.

File: master/buildbot/steps/source/git.py [hunk 1]
```diff
@@ -770,3 +661,87 @@ def _doTag(self):
...
+                 config=None, **kwargs):
+
+        self.workdir = workdir
+        self.messages = messages
+        self.files = files
+        self.env = env
+        self.timeout = timeout
+        self.logEnviron = logEnviron
+        self.config = config
+        # These attributes are not required for a git add command
```
Comment (p12tic) [inline master/buildbot/steps/source/git.py]: I think this comment could be reworded to something like: "The repourl, sshPrivateKey and sshHostKey attributes are required by GitStepMixin, but aren't needed by git add and commit operations"

File: master/buildbot/steps/source/git.py [hunk 2]
```diff
@@ -770,3 +661,87 @@ def _doTag(self):
...
+class GitCommit(buildstep.BuildStep, GitStepMixin, CompositeStepMixin):
+
+    description = None
+    descriptionDone = None
+    descriptionSuffix = None
+
+    name = 'gitcommit'
+    renderables = ['files', 'messages']
+
+    def __init__(self, workdir=None, files=None, messages=None, env=None,
```
Comment (p12tic) [inline master/buildbot/steps/source/git.py]: Are directories supported? I think better naming would be `paths`.

File: master/setup.py [hunk 3]
```diff
@@ -296,7 +296,7 @@ def define_plugin_entries(groups):
             ('buildbot.steps.source.cvs', ['CVS']),
             ('buildbot.steps.source.darcs', ['Darcs']),
             ('buildbot.steps.source.gerrit', ['Gerrit']),
-            ('buildbot.steps.source.git', ['Git', 'GitTag']),
+            ('buildbot.steps.source.git', ['Git', 'GitAdd', 'GitTag']),
```
Comment (p12tic) [inline master/setup.py]: GitCommit

**Comment by p12tic** [comment]
Looks good, thank you! I've added several comments on minor issues.

File: master/buildbot/steps/source/git.py (hunk 2 above)
Comment (chapipo) [reply master/buildbot/steps/source/git.py]: Yes, paths is a better naming.

**Comment by chapipo** [comment]
I just update the PR to take care of your comments.

I have updated docs too to reflect the merge of the 2 steps.

**Comment by p12tic** [comment]
Looks great!

**Pull request merged**